azure-proto-core/Resources/UserAssignedIdentity.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace azure_proto_core.Resources
+{
+    using System;
+    using System.Text.Json;
+    using Azure.Core;
+
+    public class UserAssignedIdentity
+    {
+        public Guid? ClientId { get; set; }
+
+        public Guid? PrincipalId { get; set; }
+
+        public UserAssignedIdentity() { }
+
+        public UserAssignedIdentity(Guid clientId, Guid principalId)
+        {
+            ClientId = clientId;
+            PrincipalId = principalId;
+        }
+
+        public int CompareTo(UserAssignedIdentity other)
+        {
+            if (other == null)
+                return 1;
+
+            if (!ClientId.HasValue && !other.ClientId.HasValue)
+                return 0;
+
+            if (!ClientId.HasValue)
+                return -1;
+
+            if (!other.ClientId.HasValue)
+                return 1;
+
+            if (ClientId.Value.CompareTo(other.ClientId.Value) == 1 && PrincipalId.Value.CompareTo(other.PrincipalId.Value) == 1)
```

Comment:
Avoid these being called 2 times by calling it once and storing its value and using it in these two cases.